File: Algorithm/Python/938. Range Sum of BST.py
# -*- coding: utf-8 -*-
# Created on Sun Nov 11 2018 10:58:0
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# simple recursive
class Solution(object):
    def rangeSumBST(self, root, L, R):
        """
        :type root: TreeNode
        :type L: int
        :type R: int
        :rtype: int
        """
        range_sum = 0
        if root == None:
            return range_sum
        if root.val <= L:
            if root.val == L:
                range_sum += root.val
            range_sum += self.rangeSumBST(root.right, L, R)
        elif root.val >= R:
            if root.val == R:
                range_sum += root.val
            range_sum += self.rangeSumBST(root.left, L, R)
        else:
            range_sum += root.val
            range_sum += self.rangeSumBST(root.left, L, root.val) + self.rangeSumBST(root.right, root.val, R)
        return range_sum

File: Algorithm/C++/938. Range Sum of BST.cpp
/*
 * Created on Mon Nov 12 2018 14:55:39
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

 // simple recursive
class Solution {
 public:
  int rangeSumBST(TreeNode* root, int L, int R) {
    if (root == NULL) return 0;
    if (root->val <= L) {
      if (root->val == L)
        return root->val + rangeSumBST(root->right, L, R);
      else
        return rangeSumBST(root->right, L, R);
    }
    else if (root->val >= R) {
      if (root->val == R)
        return root->val + rangeSumBST(root->left, L, R);
      else
        return rangeSumBST(root->left, L, R);
    }
    else {
      return root->val + rangeSumBST(root->left, L, root->val) + rangeSumBST(root->right, root->val, R);
    }
  }
}; 